README.md
# cms_project

A simple content management application that utilizes the Sinatra framework to edit delete and create files to and from filesystem

data/history.md
## Following the release of Ruby 0.95 in 1995, several stable versions of Ruby were released in the following years:

Ruby 1.0: December 25, 1996
Ruby 1.2: December 1998
Ruby 1.4: August 1999
Ruby 1.6: September 2000

In 1997, the first article about Ruby was published on the Web. In the same year, Matsumoto was hired by netlab.jp to work on Ruby as a full-time developer.

In 1998, the Ruby Application Archive was launched by Matsumoto, along with a simple English-language homepage for Ruby.

In 1999, the first English language mailing list ruby-talk began, which signaled a growing interest in the language outside Japan.[22] In this same year, Matumoto and Keiju Ishitsuka wrote the first book on Ruby, The Object-oriented Scripting Language Ruby (オブジェクト指向スクリプト言語 Ruby), which was published in Japan in October 1999. It would be followed in the early 2000s by around 20 books on Ruby published in Japanese.

By 2000, Ruby was more popular than Python in Japan. In September 2000, the first English language book Programming Ruby was printed, which was later freely released to the public, further widening the adoption of Ruby amongst English speakers. In early 2002, the English-language ruby-talk mailing list was receiving more messages than the Japanese-language ruby-list, demonstrating Ruby's increasing popularity in the non-Japanese speaking world.

test/app_test.rb
ENV["RACK_ENV"] = "test"

require 'minitest/autorun'
require 'rack/test'
require 'pry'
require "tilt/erubis"
require "fileutils"

require_relative "../cms.rb"

class AppTest < Minitest::Test
  include Rack::Test::Methods

  def setup
    FileUtils.mkdir_p(data_path)
  end

  def teardown
    FileUtils.rm_rf(data_path)
  end

  def app
    Sinatra::Application
  end

  def session
    last_request.env["rack.session"]
  end

  def create_document(name, content = "")
    File.open(File.join(data_path, name), "w") do |file|
      file.write(content)
    end
  end

  def test_index
    create_document "about.md", "about.md"
    create_document "changes.txt", "changes.txt"

    get "/"

    assert_equal 200, last_response.status
    assert_equal "text/html;charset=utf-8", last_response["Content-Type"]
    assert_includes last_response.body, "about.md"
    assert_includes last_response.body, "changes.txt"
  end

  def test_about_page
    create_document "about.md", "Early Concept"
    get "/about.md"

    assert_includes(last_response.body, "Early Concept")
  end

  def test_nonexistant_route
    get "/not_a_file.md"
    redirected_request = get "/"
    binding.pry
    assert_includes(redirected_request.body, "not_a_file.md does not exist")
  end

  def test_edit_document
    create_document "about.md"

    post_res = post "/about.md/edit_file"
    get_redirect_res = get "/"

    assert_includes get_redirect_res.body, "about.md has been updated"
    assert_equal post_res.status, 302
  end

  def test_new_document_creation
    post_res = post "/new_file?file_name=this.md"
    redirect_res = get "/"

    assert_includes redirect_res.body, "File: this.md has been created"
    assert_equal post_res.status, 302
  end

  def test_new_document_fail
    post_res = post "/new_file?file_name="

    assert_equal post_res.status, 422
    assert_includes post_res.body, "The file name must be provided."
  end

  def test_delete_document
    create_document "document.md"
    post_res = post "/delete/document.md"
    redirect_res = get "/"

    assert_includes redirect_res.body, "document.md has been deleted"
    assert_equal 302, post_res.status
  end

  def test_delete_fail
    post_res = post "/delete/document.notreal"

    assert_equal 422, post_res.status
    assert_includes post_res.body, "The file could not be found to be deleted"
  end
end

cms.rb
require "sinatra"
require "sinatra/reloader" if development?
require "sinatra/content_for"
require "tilt/erubis"
require "pry"
require "find"
require "redcarpet"

configure do
  enable :sessions
  set :session_secret, 'super secret'
end

def data_path
  if ENV["RACK_ENV"] == "test"
    File.expand_path("../test/data", __FILE__)
  else
    File.expand_path("../data", __FILE__)
  end
end

def render_markdown(text)
  markdown = Redcarpet::Markdown.new(Redcarpet::Render::HTML)
  markdown.render(text)
end

before do
  @root = File.expand_path("..", __FILE__)
  file_path = File.join data_path, "*"

  @files = Dir.glob(file_path).map do |path|
             File.basename(path)
           end
end

get "/" do
  pattern = File.join(data_path, "*")
  erb :files
end

def load_file_content(path)
  content = File.read(path)
  case File.extname(path)
  when ".txt"
    headers["Content-Type"] = "text/plain"
    content
  when ".md"
    erb render_markdown(content)
  end
end

get "/new_file" do
  erb :new_file
end

get "/sign_in" do
  erb :sign_in
end

get "/:file_name" do
  @file_name = params['file_name']
  file_path = File.join(data_path, @file_name)
  session[:hello] = "whatup"
  
  if @files.include? @file_name
    load_file_content file_path
  else
    session[:error] = "#{@file_name} does not exist"
    redirect "/"
  end
end

get "/:file_name/edit_file" do
  @file_name = params['file_name']
  file_path = File.join(data_path, @file_name)

  if @files.include? @file_name
    @file = File.read(file_path)
    erb :edit_file
  else
    session[:error] = "#{@file_name} does not exist"
    redirect "/"
  end
end

post "/:file_name/edit_file" do
  file_name = params[:file_name]
  file_path = File.join(data_path, file_name)

  file = File.open(file_path , "w") do |f|
    f.write params[:fileContents]
  end

  session[:success] = "#{params[:file_name]} has been updated"
  redirect "/"
end

post "/new_file" do
  file_name = params[:file_name]
  file_path = File.join(data_path, file_name)
  
  unless file_name.empty?
    File.write(file_path, "")
    session[:success] = "File: #{params[:file_name]} has been created"
    status 302
    redirect "/"
  else
    session[:error] = "The file name must be provided."
    status 422
    erb :new_file
  end
end

post "/delete/:file_name" do
  file_name = params[:file_name]
  path = data_path
  if @files.include? file_name
    File.delete(File.join(path, file_name))
    session[:success] = "#{file_name} has been deleted."
    redirect "/"
  else
    status 422
    session[:error] = "The file could not be found to be deleted"
    erb :files
  end
end

post "/sign_in" do
  session[:username] = params[:username]
  @username = session[:username]
  password = params[:password]

  if session[:username] == 'admin' && password == 'secret'
    session[:signed_in] = true
    session[:success] = "Welcome!"
    redirect "/"
  else
    session[:error] = "Wrong username or password"
    erb :sign_in
  end
end

post "/sign_out" do
  session[:signed_in] = false
  session[:success] = "You have been signed out."
  redirect "/"
end
